### PR TITLE
feat: display percent in progress bar for web client

### PR DESCRIPTION
**Pull request by Rukario**
<details>
<summary>Old screenshot</summary>

Transmission percent 1

Animated diff from original
Transmission percent 1 (apng)</details>

Notes: Added percent digits into the progress bar.

### CODE REVIEW

**Comment by Rukario** [comment]
Requesting design & code review from @dareiff

**Comment by GaryElshaw** [comment]
Or add this pr to #5857?

**Comment by ckerr** [comment]
@Rukario sorry for leaving this in code review limbo for so long, I'd like to get this PR unstuck. Could you update this & resolve the scss conflicts?

**Comment by Rukario** [comment]
@ckerr I can resolve the file conflict but there's more things this PR need to cover, particularly during contrast mode where percent will become invisible on white bar whereas non-contrast would have it light gray background, I'll mark this PR WIP for now.

**Comment by Rukario** [comment]
It's been a while I've touched my developer setup for Transmission. I'll catch up and I'll be sure to update this PR when everything is in working order with the latest Transmission.

**Comment by Rukario** [comment]
Left: this PR, right: main

Transmission percent light
Transmission percent light contrast
Transmission percent dark
Transmission percent dark contrast

The percent text became invisible upon white background of the bar under high contrast for seeding, I revised it to use the green progress bar from non-high contrast and then use black background to resolve this.

Any suggestion for the queued progress bar that the new percent text is hard to see there? My custom client has same blue color for both queued and downloading, how does that sound?

**Comment by Rukario** [comment]
I hope everything's fine.

**Comment by Pentaphon** [comment]
Any update on this? This one has been stuck for a long time.

**Review by tearfur — CHANGES_REQUESTED** [review]
I will only review in terms of code quality, I don't have much opinion about the idea itself. 

I'd like these changes below applied (preferrably with me as commit author `Yat Ho <lagoho7@gmail.com>`). They mainly reduce code duplication, and makes better use of SCSS features.

```patch
diff --git a/web/assets/css/transmission-app.scss b/web/assets/css/transmission-app.scss
index 3ceea6f74..41346b450 100644
--- a/web/assets/css/transmission-app.scss
+++ b/web/assets/css/transmission-app.scss
@@ -106,9 +106,9 @@
     --color-fg-selected: var(--dark-mode-white);
     --color-fg-tertiary: var(--grey-500);
     --color-fg-warn: var(--dark-mode-black);
+    --color-progressbar-fg-1: #edefff;
+    --color-progressbar-fg-2: #edefff;
     --color-progressbar-paused: var(--grey-500);
-    --color-progressbar-paused-fg: #edefff;
-    --color-progressbar-paused-fg-2: #edefff;
     --color-progressbar-seed-1: var(--green-100);
     --color-progressbar-seed-2: var(--green-400);
     --color-progressbar-seed-paused: var(--grey-500);
@@ -136,10 +136,10 @@
       --color-fg-tertiary: var(--white);
       --color-fg-warn: var(--black);
       --color-progressbar-background-2: var(--white);
+      --color-progressbar-fg-1: #fff;
+      --color-progressbar-fg-2: #fff;
       --color-progressbar-magnet: var(--yellow-300);
       --color-progressbar-paused: var(--grey-500);
-      --color-progressbar-paused-fg: #fff;
-      --color-progressbar-paused-fg-2: #fff;
       --color-progressbar-queued: var(--blue-400);
       --color-progressbar-seed-2: var(--black);
       --color-progressbar-seed-paused: var(--grey-500);
@@ -168,11 +168,11 @@
     --color-fg-selected: var(--nice-grey);
     --color-fg-tertiary: var(--grey-500);
     --color-fg-warn: #cf212e;
+    --color-progressbar-fg-1: #303030;
+    --color-progressbar-fg-2: #edefff;
     --color-progressbar-leech: var(--blue-100);
     --color-progressbar-magnet: var(--yellow-300);
     --color-progressbar-paused: var(--grey-200);
-    --color-progressbar-paused-fg: #303030;
-    --color-progressbar-paused-fg-2: #edefff;
     --color-progressbar-queued: var(--blue-400);
     --color-progressbar-seed-1: var(--green-500);
     --color-progressbar-seed-2: var(--green-300);
@@ -202,11 +202,11 @@
       --color-fg-tertiary: var(--black);
       --color-fg-warn: var(--white);
       --color-progressbar-background-2: var(--white);
+      --color-progressbar-fg-1: #fff;
+      --color-progressbar-fg-2: #fff;
       --color-progressbar-leech: var(--blue-200);
       --color-progressbar-magnet: var(--yellow-300);
       --color-progressbar-paused: var(--grey-500);
-      --color-progressbar-paused-fg: #fff;
-      --color-progressbar-paused-fg-2: #fff;
       --color-progressbar-queued: var(--blue-400);
       --color-progressbar-seed-2: var(--black);
       --color-progressbar-seed-paused: var(--grey-500);
@@ -699,12 +699,12 @@ a {
       &.queued::before {
         background: var(--color-progressbar-queued);
       }
-    }
 
-    &.leech::before {
-      background: var(--color-progressbar-leech);
-      color: var(--color-progressbar-paused-fg-2);
-      text-shadow: var(--progress-bar-shadow-2);
+      &::before {
+        background: var(--color-progressbar-leech);
+        color: var(--color-progressbar-fg-2);
+        text-shadow: var(--progress-bar-shadow-2);
+      }
     }
 
     &.magnet::before {
@@ -715,31 +715,33 @@ a {
     &.seed {
       &.paused::before {
         background: var(--color-progressbar-seed-paused);
-        color: var(--color-progressbar-paused-fg);
+        color: var(--color-progressbar-fg-1);
         text-shadow: var(--progress-bar-shadow);
       }
 
-      &:not(.paused)::before {
-        background: var(--color-progressbar-seed-1);
-        color: var(--color-progressbar-paused-fg-2);
-        text-shadow: var(--progress-bar-shadow-2);
-      }
+      &:not(.paused) {
+        &::before {
+          background: var(--color-progressbar-seed-1);
+          color: var(--color-progressbar-fg-2);
+          text-shadow: var(--progress-bar-shadow-2);
+        }
 
-      &:not(.paused)::after {
-        background: var(--color-progressbar-seed-2);
-        color: var(--color-progressbar-paused-fg-2);
-        text-shadow: var(--progress-bar-shadow-2);
+        &::after {
+          background: var(--color-progressbar-seed-2);
+          color: var(--color-progressbar-fg-2);
+          text-shadow: var(--progress-bar-shadow-2);
+        }
       }
 
       &.queued::before {
         background-color: var(--color-progressbar-seed-1);
       }
-    }
 
-    &.seed::before {
-      background-color: var(--color-progressbar-seed-1);
-      color: var(--color-progressbar-paused-fg-2);
-      text-shadow: var(--progress-bar-shadow-2);
+      &::before {
+        background-color: var(--color-progressbar-seed-1);
+        color: var(--color-progressbar-fg-2);
+        text-shadow: var(--progress-bar-shadow-2);
+      }
     }
 
     &.verify::before {
@@ -749,37 +751,39 @@ a {
 
     &.paused::before {
       background: var(--color-progressbar-paused);
-      color: var(--color-progressbar-paused-fg);
+      color: var(--color-progressbar-fg-1);
       text-shadow: var(--progress-bar-shadow);
     }
   }
 
-  .torrent-progress-bar::before,
-  .torrent-progress-bar::after {
-    content: attr(data-progress);
-    height: 100%;
-    width: 100%;
-    position: absolute;
-    border-radius: 2px;
-    text-align: center;
-  }
-
-  .torrent-progress-bar::before {
-    clip-path: polygon(
-      0 0,
-      var(--progress, 30%) 0,
-      var(--progress, 30%) 100%,
-      0 100%
-    );
-  }
-
-  .torrent-progress-bar::after {
-    clip-path: polygon(
-      var(--progress, 30%) 0,
-      100% 0,
-      100% 100%,
-      var(--progress, 30%) 100%
-    );
+  .torrent-progress-bar {
+    &::before,
+    &::after {
+      content: attr(data-progress);
+      height: 100%;
+      width: 100%;
+      position: absolute;
+      border-radius: 2px;
+      text-align: center;
+    }
+
+    &::before {
+      clip-path: polygon(
+        0 0,
+        var(--progress, 30%) 0,
+        var(--progress, 30%) 100%,
+        0 100%
+      );
+    }
+
+    &::after {
+      clip-path: polygon(
+        var(--progress, 30%) 0,
+        100% 0,
+        100% 100%,
+        var(--progress, 30%) 100%
+      );
+    }
   }
 }
 
diff --git a/web/src/file-row.js b/web/src/file-row.js
index 3d9885b6e..d378a3848 100644
--- a/web/src/file-row.js
+++ b/web/src/file-row.js
@@ -34,6 +34,7 @@ export class FileRow extends EventTarget {
     const fmt = Formatter;
     const c = `${fmt.size(have)} of ${fmt.size(size)} (${fmt.percentString(
       pct,
+      1,
     )}%)`;
     setTextContent(this.elements.progress, c);
   }
diff --git a/web/src/formatter.js b/web/src/formatter.js
index 105ddb521..7389ae908 100644
--- a/web/src/formatter.js
+++ b/web/src/formatter.js
@@ -94,8 +94,8 @@ export const Formatter = {
   },
 
   // format a percentage to a string
-  percentString(x) {
-    const decimal_places = x < 100 ? 1 : 0;
+  percentString(x, decimal_places) {
+    decimal_places = x < 100 ? decimal_places : 0;
     return this._toTruncFixed(x, decimal_places);
   },
 
@@ -109,7 +109,7 @@ export const Formatter = {
     if (x === -2) {
       return '&infin;';
     }
-    return this.percentString(x);
+    return this.percentString(x, 1);
   },
 
   /**
diff --git a/web/src/inspector.js b/web/src/inspector.js
index ed662200b..ed24e0f04 100644
--- a/web/src/inspector.js
+++ b/web/src/inspector.js
@@ -292,7 +292,7 @@ export class Inspector extends EventTarget {
       const d =
         100 *
         (sizeWhenDone ? (sizeWhenDone - leftUntilDone) / sizeWhenDone : 1);
-      string = fmt.percentString(d);
+      string = fmt.percentString(d, 1);
 
       if (unverified) {
         string = `${fmt.size(verified)} of ${fmt.size(
@@ -319,7 +319,7 @@ export class Inspector extends EventTarget {
         (accumulator, t) => t.getHave() + t.getDesiredAvailable(),
         0,
       );
-      string = `${fmt.percentString((100 * available) / sizeWhenDone)}%`;
+      string = `${fmt.percentString((100 * available) / sizeWhenDone, 1)}%`;
     }
     setTextContent(e.info.availability, fmt.stringSanitizer(string));
 
diff --git a/web/src/torrent-row.js b/web/src/torrent-row.js
index 5577b6253..ffd36235c 100644
--- a/web/src/torrent-row.js
+++ b/web/src/torrent-row.js
@@ -80,11 +80,10 @@ const TorrentRendererHelper = {
 
   renderProgressbar: (controller, t, progressbar) => {
     const info = TorrentRendererHelper.getProgressInfo(controller, t);
+    const pct_str = `${Formatter.percentString(info.percent, 2)}%`;
     progressbar.className = info.classList.join(' ');
-    progressbar.style.setProperty('--progress', `${info.percent.toFixed(2)}%`);
-    progressbar.dataset.progress = `${Number.parseFloat(
-      info.percent.toFixed(2),
-    )}%`;
+    progressbar.style.setProperty('--progress', pct_str);
+    progressbar.dataset.progress = pct_str;
   },
 };
 
@@ -154,7 +153,7 @@ export class TorrentRendererFull {
     if (t.isChecking()) {
       return [
         'Verifying local data (',
-        Formatter.percentString(100 * t.getRecheckProgress()),
+        Formatter.percentString(100 * t.getRecheckProgress(), 1),
         '% tested)',
       ].join('');
     }
@@ -171,7 +170,7 @@ export class TorrentRendererFull {
       const percent = 100 * t.getMetadataPercentComplete();
       return [
         `Magnetized transfer - ${MetaDataStatus} metadata (`,
-        Formatter.percentString(percent),
+        Formatter.percentString(percent, 1),
         '%)',
       ].join('');
     }
diff --git a/web/src/torrent.js b/web/src/torrent.js
index 404dafd83..2fa15071f 100644
--- a/web/src/torrent.js
+++ b/web/src/torrent.js
@@ -280,7 +280,7 @@ export class Torrent extends EventTarget {
     return this.getDownloadSpeed() + this.getUploadSpeed();
   }
   getPercentDoneStr() {
-    return Formatter.percentString(100 * this.getPercentDone());
+    return Formatter.percentString(100 * this.getPercentDone(), 1);
   }
   getPercentDone() {
     return this.fields.percentDone;
```

**Comment by Rukario** [comment]
@tearfur I like your changes, it cleaned up the code a bit. I like how the paused progress bar is supposed to be the default colorless bar. That really gives different perspective.

I've applied changes manually and learned today how to co-author someone on GitHub.

My transmission setup exploded with new files, waiting for PR to finally ship so I can clean up the mess. Appreciate your feedback but I'll have to keep files maintainable. I don't have expertise with git stuff and I'm sure I don't do it quite correctly.

**Review by tearfur — APPROVED** [review]
I'll leave the UI changes (particularly the colour changes in high contrast mode) for others to judge.

The code looks good to me. 👍🏻

**Comment by tearfur** [comment]
> I like how the paused progress bar is supposed to be the default colorless bar. That really gives different perspective.

Err, what are you referring to with this? I thought my suggested changes don't modify any behaviour.

**Comment by Pentaphon** [comment]
> I'll leave the UI changes (particularly the colour changes in high contrast mode) for others to judge.

UI changes look good to me. Any edge cases most people don't see due to variances in sight ability can be dealt with later with color tweaks. I don't see any glaring legibility issues.

**Comment by Rukario** [comment]
> > I like how the paused progress bar is supposed to be the default colorless bar. That really gives different perspective.
> 
> Err, what are you referring to by this? I thought my suggested changes don't modify any behaviour.

No changes visually don't worry about it, what I mean is a change in variable name will make you see different things.

File: web/assets/css/transmission-app.scss [hunk 1]
```diff
@@ -136,9 +138,10 @@
       --color-progressbar-background-2: var(--white);
       --color-progressbar-magnet: var(--yellow-300);
       --color-progressbar-paused: var(--grey-500);
+      --color-progressbar-paused-fg: #fff;
+      --color-progressbar-paused-fg-2: #fff;
       --color-progressbar-queued: var(--blue-400);
-      --color-progressbar-seed-1: var(--black);
```
[the remark this answers is not in the document]
Comment (Rukario) [reply web/assets/css/transmission-app.scss:140]: The percent text on bright green bar of non-contrast is borderline illegible, I think color blind will appreciate black background more in contrast mode.

[the remark this answers is not in the document]
Comment (tearfur) [reply web/assets/css/transmission-app.scss:140]: Then how about keeping the original black colour, then use white fg?

[the remark this answers is not in the document]
Comment (Rukario) [reply web/assets/css/transmission-app.scss:140]: Actually could try, it was last-minute change for percent text on bar so I don't get to add more variables for black/white in contrast mode. Should we?

[the remark this answers is not in the document]
Comment (tearfur) [reply web/assets/css/transmission-app.scss:140]: I say why not, should be very simple.

**Comment by Rukario** [comment]
Left: This PR, Right: Original
Transmission percent light
Transmission percent light contrast
Transmission percent dark
Transmission percent dark contrast

File: web/assets/css/transmission-app.scss (hunk 1 above)
[the remark this answers is not in the document]
Comment (Rukario) [reply web/assets/css/transmission-app.scss]: I'll have to question on the color choices for the bars in high contrast modes but I know this PR shouldn't be stopped by it. Just a few tweaks away so yeah why not.

**Review by tearfur — APPROVED** [review]

File: web/assets/css/transmission-app.scss [hunk 2]
```diff
@@ -684,73 +706,93 @@ a {
...
-        ),
-        no-repeat;
+    &.paused::before {
+      background: var(--color-progressbar-paused);
+      color: var(--color-progressbar-fg-1);
+      text-shadow: var(--progress-bar-shadow-1);
+    }
+  }
+
+  .torrent-progress-bar {
```
Comment (tearfur) [inline web/assets/css/transmission-app.scss:769]: These 2 `.torrent-progress-bar` blocks can be merged.

Comment (Rukario) [reply web/assets/css/transmission-app.scss]: Thank you

**Comment by Rukario** [comment]
Hope to see this merged into the main soon,

**Review by ckerr — APPROVED** [review]

**Comment by tearfur** [comment]
@Rukario I'm brainstorming of methods to collapse the `*-1` `*-2` `*-3` variables to functions. I have a question for you.

So far I have an idea for the fg, but are stumped at the text shadows.

Do you have a systematic criteria for when and when not to have shadows? I want to figure out what the criteria is, and whether it can be expressed in code.

**Comment by Rukario** [comment]
@tearfur Yes, it's the white text that would get shadow. I can see several places in the CSS that could be simplified and maybe cut back some unnecessary styles. You're welcome to make a PR for that.

**Pull request merged**